---
title: "Using haploR, an R package for querying HaploReg and RegulomeDB"
author: "Ilya Y. Zhbannikov"
date: '`r date()`'
output:
  html_document: default
  #pdf_document: default
  #md_document : 
  #  variant: markdown_github
  
vignette: >
  %\VignetteIndexEntry{Tutorial for haploR package} 
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---


## Overview

HaploReg (<http://archive.broadinstitute.org/mammals/haploreg/haploreg.php>) and 
RegulomeDB (<http://www.regulomedb.org>)
are web-based tools that extract biological information such as eQTL, 
LD, motifs, etc. from large genomic projects such as ENCODE, 
the 1000 Genomes Project, Roadmap Epigenomics Project and others. 
This is sometimes called "post stage GWAS" analysis.

The R-package _haploR_ was developed to query 
those tools (HaploReg and RegulomeDB) directly from 
_R_ in order to facilitate high-throughput 
genomic data analysis. Below we provide several examples 
that show how to work with this package.

Note: you must have a stable Internet connection 
to use this package.

Contact: <ilya.zhbannikov@duke.edu> for questions of 
usage the _haploR_ or any other issues.

### Motivation and typical analysis workflow

This package was inspired by the fact that many web-based post stage GWAS databases do not have Application Programing Interface (API) and, therefore, do not allow users to query them remotedly from R environment. In our research we used HaploReg and RegulomeDB web databases. These very useful web databases show information about linkage disequilibrium of query variants and variants which are in LD with them, expression quantitative trait loci (eQTL), motifs changed and other useful information. However, it is not easy to include this information into streamlined workflow since those tools also not offer API.

We developed a custom analysis pipeline which prepares data, performs genome-wide association (GWA) analysis and presents results in a user-friendly format. Results include a list of genetic variants (also known as 'SNP' or single nucleotide polymorphism), their corresponding _p_-values, phenotypes (traits) tested and other meta-information such as LD, alternative allele, minor allele frequency, motifs changed, etc. Of course, we could go throught the list of SNPs having genome-wide significant _p_-values (1e-8) and submit each of those SNPs to HaploReg or RegulomeDB manually, one-by-one, but it is time-consuming process and will not be fully automatic (which ruins one of the pipeline's paradigms). This is especially difficult if the web site does not offer downloading results.

Therefore, we developed _haploR_, a user-friendly R package that connects to HaploReg or RegulomeDB from R environment with methods POST and GET and downloads results in a suitable R format. This package siginificantly saved our time in developing reporting system for our internal genomic analysis pipeline and now we would like to present _haploR_ to the research community.

Example of typical analysis workflow is shown below.

![Typical analysis workflow](Workflow.png)

* Data preprocessing stage usually consists of basic cleaning operations (sex inconsistencies check, filtering by MAF, missing genotype rate), format conversion, population stratification, creating temporaty files, etc.
* Genome-wide association study (GWAS). This Includes testing hyphotesis on correlation between phenotype and genotype. Usually in form of linear or logistic regression but can be quite sophisticated especially for rare variants.
* Postprocessing usually includes summarizing results in tables, creating graphics (Manhattan plot, QQ-plot), sometimes filtering by significance threshold (usually 1E-8), removing temporary files, etc.
* Post stage GWAS analysis: connect GWAS findings with existing studies and gather information about linked SNPs, chromatin state, protein binding annotation, sequence conservation across mammals, effects on regulatory motifs and on gene expression, etc. This helps researchers to more understand functions and test additional hyphotesis of found genome-wide significant SNPs. At this stage _haploR_ is especially useful because it provides a convenient R interface to mining web databases. This, in turn, streamlines analysis workflow and therefore significantly reduces analysis time. Previously researchers had to do it manually after analysis by exploring available web resources, checking each SNP of interest and downloading results (which is especially painful if website does not have a download option). With _haploR_ such post-GWAS information are easily retrieved, shared and appeared in the final output tables at the end of analysis. This will save researchers' time.

## Installation of _haploR_ package

In order to install the _haploR_ package, the user must first install R (<https://www.r-project.org>). After that, _haploR_ can be installed either:

* From CRAN (stable version):
```{r, echo=TRUE, eval=FALSE}
install.packages("haploR", dependencies = TRUE)
```

* Or from the package web page (developing version):

```{r, echo=TRUE, eval=FALSE}
devtools::install_github("izhbannikov/haplor")
```


_haploR_ depends on the following packages: 

* _httr_, version 1.2.1 or later.
* _XML_, version version 3.98-1.6 or later.
* _tibble_, version 1.3.0 or later.
* _RUnit_, version 0.4.31 or later.

## Examples of usage

### Querying HaploReg

Function 
```
queryHaploreg(query = NULL, file = NULL, study = NULL, ldThresh = 0.8,
  ldPop = "EUR", epi = "vanilla", cons = "siphy", genetypes = "gencode",
  url = "http://archive.broadinstitute.org/mammals/haploreg/haploreg.php",
  timeout = 10, encoding = "UTF-8", verbose = FALSE)
```

queries HaploReg web-based tool and returns results.

#### Arguments

* _query_:	Query (a vector of rsIDs).
* _file_:	A text file (one refSNP ID per line).
* _study_:	A particular study. See function getHaploRegStudyList(...). Default: ```NULL```.
* _ldThresh_:	LD threshold, r2 (select NA to show only query variants). Default: ```0.8```.
* _ldPop_:	1000G Phase 1 population for LD calculation. Can be: ```AFR``` (Africa), ```AMR``` (America), ```ASN``` (Asia). Default: ```EUR``` (Europe).
* _epi_: Source for epigenomes. Possible values: ```vanilla``` for ChromHMM (Core 15-state model); ```imputed``` for ChromHMM (25-state model using 12 imputed marks); ```methyl``` for H3K4me1/H3K4me3 peaks; ```acetyl``` for H3K27ac/H3K9ac peaks. Default: ```vanilla```.
* _cons_:	Mammalian conservation algorithm. Possible values: ```gerp``` for GERP (<http://mendel.stanford.edu/SidowLab/downloads/gerp/>), ```siphy``` for SiPhy-omega, ```both``` for both. Default: siphy.
* _genetypes_:	Show position relative to. Possible values: ```gencode``` for Gencode genes; ```refseq``` for RefSeq genes; ```both``` for both. Default: ```gencode```.
* _url_: HaploReg url address. Default: <http://archive.broadinstitute.org/mammals/haploreg/haploreg.php>
* _timeout_:	A timeout parameter for curl. Default: ```10```
* _encoding_: Set the encoding for correct retrieval web-page content. Default: ```UTF-8```
* _verbose_: Verbosing output. Default: ```FALSE```.

#### Output

A data frame (table) wrapped into a _tibble_ object contains data extracted from HaploReg site. The colums (33 in total at the time of writing this vignette) are specific to HaploReg output. Below we describe the columns:

* _chr_: Chromosome, type: numeric
* _pos_hg38_: Position on the human genome, type: numeric.
* _r2_: Linkage disequilibrium. Type: numeric.
* D': Linkage disequilibrium, alternative definition. Type: numeric.
* _is_query_snp_: Indicator shows query SNP, 0 - not query SNP, 1 - query SNP. Type: numeric.
* _rsID_: refSNP ID. Type: character.
* _ref_: Reference allele. Type: character.
* _alt_: Alternative allele. Type: character.
* _AFR_: _r2_ calculated for Africa. Type: numeric.
* _AMR_: _r2_ calculated for America. Type: numeric.
* _ASN_: _r2_ calculated for Asia. Type: numeric.
* _EUR_: _r2_ calculated for Europe. Type: numeric.
* _GERP_cons_: GERP scores. Type: numeric.
* _SiPhy_cons_: SiPhy scores. Type: numeric.
* _Chromatin_States_: Chromatin states: reference epigenome identifiers (EID) of chromatin-associated proteins and histone modifications in that region. Type: character.
* _Chromatin_States_Imputed_: Chromatin states based on imputed data. Type: character.
* _Chromatin_Marks_: Chromatin marks Type: character.
* _DNAse_: DNAse. Type: chararcter.
* _Proteins_: A list of protein names. Type: character.
* _eQTL_: Expression Quantitative Trait Loci. Type: character.
* _gwas_: GWAS study name. Type: character.
* _grasp_: GRASP study name: character.
* _Motifs_: Motif names. Type: character.
* _GENCODE_id_: GENCODE transcript ID. Type: character.
* _GENCODE_name_: GENCODE gene name. Type: character.
* _GENCODE_direction_: GENCODE direction (transcription toward 3' or 5' end of the DNA sequence). Type: numeric.
* _GENCODE_distance_: GENCODE distance. Type: numeric.
* _RefSeq_id_: NCBI Reference Sequence Accession number. Type: character.
* _RefSeq_name_: NCBI Reference Sequence name. Type: character.
* _RefSeq_direction_: NCBI Reference Sequence direction (transcription toward 3' or 5' end of the DNA sequence). Type: numeric.
* _RefSeq_distance_: NCBI Reference Sequence distance. Type: numeric.
* _dbSNP_functional_annotation_ Annotated proteins associated with the SNP. Type: numeric. 
* _query_snp_rsid_: Query SNP rs ID. Type: character.

Number of rows is not constant, at least equal or more than the number of query SNPs, and depends on _r2_ parameter choosen in a query (default 0.8). This means that the program outputs not only query SNPs, but also those SNPs that have _r2_ $\geq$ 0.8 with the query SNPs.

#### One or several genetic variants

```{r, echo=TRUE, message=FALSE}
library(haploR)
x <- queryHaploreg(query=c("rs10048158","rs4791078"))
x
```

Here _query_ is a vector with names of genetic variants. 

We then can create a subset from the results, for example, to choose only SNPs with r2 $>$ 0.9:

```{r, echo=TRUE, message=FALSE}
subset.high.LD <- x[x$r2 > 0.9, c("rsID", "r2", "chr", "pos_hg38", "is_query_snp", "ref", "alt")]
subset.high.LD
```

We can then save the _subset.high.LD_ into an Excel workbook:

```{r, echo=TRUE, message=FALSE, eval=FALSE}
require(openxlsx)
write.xlsx(x=subset.high.LD, file="subset.high.LD.xlsx")
```


This was an example of gathering post-gwas information directly from the online tool. _haploR_ has an additional advantage because it downloads the full information for query retrieved by HaploReg. For example, if you go online and submit these two SNPs to HaploReg (<http://archive.broadinstitute.org/mammals/haploreg/haploreg.php>), you will see that some cells of columns "Motifs changed" and "Selected eQTL hits" are hidded (only number of hits are given). _haploR_ retrives this information in a form of a data frame which can be saved into Excel file.

```{r, echo=TRUE, message=FALSE}
x[, c("Motifs", "rsID")]
x[, c("eQTL", "rsID")]
```

#### Uploading file with variants

If you have a file with your SNPs you would like 
to analyze, you can 
supply it as an input as follows:

```{r, echo=TRUE, message=FALSE}
library(haploR)
x <- queryHaploreg(file=system.file("extdata/snps.txt", package = "haploR"))
x
```

File "snps.txt" is a text file which contains one rs-ID per line:

```
rs10048158
rs4791078
```

#### Using existing studies

Sometimes one would like to explore results from 
already performed study. In this case you should 
first explore the existing studies from 
HaploReg web site (<http://archive.broadinstitute.org/mammals/haploreg/haploreg.php>) and then use one of 
them as an input parameter. See example below:

```{r, echo=TRUE, message=FALSE}
library(haploR)
# Getting a list of existing studies:
studies <- getStudyList()
# Let us look at the first element:
studies[[1]]
# Let us look at the second element:
studies[[2]]
# Query Hploreg to explore results from 
# this study:
x <- queryHaploreg(study=studies[[1]])
x
```

### Querying RegulomeDB

To query RegulomeDB use this function:
```
queryRegulome(query = NULL, 
              format = "full",
              url = "http://www.regulomedb.org/results", 
              timeout = 10,
              check_bad_snps = TRUE, 
              verbose = FALSE)
```

This function queries RegulomeDB <http://www.regulomedb.org> web-based tool and returns results in a named list.

#### Arguments

* _query_: Query (a vector of rsIDs).
* _format_: An output format.  Only 'full' is currently supported. See ```http://www.regulomedb.org/results```. Can be on of the following: ```full``` - plain text, ```bed``` -  BED (Browser Extensible Data) format, see e.g. <https://genome.ucsc.edu/FAQ/FAQformat.html#format5.1>, ```gff``` - GFF (General Feature Format), see e.g. <https://genome.ucsc.edu/FAQ/FAQformat.html#format3>. Only ```full``` is currently supported. 
* _url_: Regulome url address.  Default: ```http://www.regulomedb.org/results```
* _timeout_: A 'timeout' parameter for 'curl'. Default: 10.
* _check_bad_snps_: Checks if all query SNPs are annotated (i.e. presented in the Regulome Database). Default: 'TRUE'
* _verbose_: Verbosing output. Default: FALSE.

#### Output

A list of two: (1) a data frame (res.table) wrapped to a _tibble_ object and (2) a list of bad SNP IDs (bad.snp.id).  Bad SNP ID are those IDs that were not found in 1000 Genomes Phase 1 data and, therefore, in RegulomeDB.

Columns in a data frame (res.table):

* _#chromosome_: Chromosome. Type: character.
* _coordinate_: Position. Type: numeric.
* _rsid_: RefSeq SNP ID. Type: character.
* _hits_: Contains information about chromatin structure: method and cell type. Type: character.
* _score_: Internal RegulomeDB score. See <http://www.regulomedb.org/help#score>. Type: numeric.

Number of rows is equal or less (in case if not all SNPs were found in RegulomeDB database) to number of query SNPs.

All the information retreived from RegulomeDB, except _hits_, are also presented in HaploReg output (```queryHaploReg(...)```).

#### Example

```{r, echo=TRUE, message=FALSE}
library(haploR)
x <- queryRegulome(c("rs4791078","rs10048158"))
x$res.table
x$bad.snp.id
```



## Session information
```{r, echo=TRUE}
sessionInfo()
```
